### PR TITLE
Fix #167, #202, #204, #274 and a bunch of other goodies

### LINKED ISSUE #167: Empty string arguments not being passed to git
Currently, the `git()` function defined in `aliases.sh` does not retain empty string arguments. Here's an example to illustrate:

``` bash
$ /usr/local/bin/git commit --allow-empty --allow-empty-message -m ''
[master f2a34ac]

$ /usr/local/bin/hub commit --allow-empty --allow-empty-message -m ''
[master e649005]

$ git commit --allow-empty --allow-empty-message -m ''
error: switch `m' requires a value
usage: git commit [<options>] [--] <pathspec>...
...
```

Related: github/hub#727

### LINKED ISSUE #183: Breaks cd when using other languages
I am using ubuntu 14.04 (in portuguese).

After install scm_breeze and sourcing it in my .bashrc, when I try to change to a dir I get:

```
user@mycomputer:~$ cd tdc/
bash: tdc/: É um diretório
```

(tdc: Is a directory)

When I change my system settings to use english, everything seems to work fine.

Running this cd in `bash -X`, i get:

```
user@mycomputer:~$ cd tdc/
+ exec_scmb_expand_args tdc/
++ scmb_expand_args tdc/
++ '[' tdc/ = --relative ']'
++ first=1
++ OLDIFS='     
'
++ IFS=' '
++ for arg in '"$@"'
++ [[ tdc/ =~ ^[0-9]{0,4}$ ]]
++ [[ tdc/ =~ ^[0-9]+-[0-9]+$ ]]
++ '[' 1 -eq 1 ']'
++ first=0
++ printf %s tdc/
++ IFS='    
'
++ sed -e 's/\([][|;()<>^ "'\'']\)/\\\1/g'
+ eval tdc/
++ tdc/
bash: tdc/: É um diretório
++ history -a
```

I am using the stock settings of scm_breeze, also I reset my .bashrc to stock settings.

### LINKED ISSUE #202: Wrong argument quoting when filename uses spaces.
hi @ndbroadbent, I seem to be getting some quoting errors when using the aliased `mv` using filenames with spaces, here's an example (with bash)

```
> mv Tungevaag\ \&\ Raaban\,\ Emila\ -Samsara.mp3 Tungevaag\ \&\ Raaban\,\ Emila\ -\ Samsara.mp3                  
[1] 30313
[2] 30314
/bin/mv: missing destination file operand after ‘Tungevaag ’
Try '/bin/mv --help' for more information.
 Raaban, Emila -Samsara.mp3: command not found
 Raaban, Emila - Samsara.mp3: command not found
[1]-  Exit 1                  /bin/mv Tungevaag\ 
[2]+  Exit 127                \ Raaban,\ Emila\ -Samsara.mp3 Tungevaag\ 
```

```
> type mv
mv is aliased to `exec_scmb_expand_args /bin/mv'

> type exec_scmb_expand_args
exec_scmb_expand_args is a function
exec_scmb_expand_args () 
{ 
    eval "$(scmb_expand_args "$@" | sed -e "s/\([][|;()<>^ \"']\)/"'\\\1/g')"
}
```

I think the right way to deal with would be to use something along the lines of `printf "%q" "$variable"`.
I can work on that, would you be accepting a MR using this strategy?

It is also possible that such a patch would fix another couple of issues that have already been reported.

### LINKED ISSUE #204: File names with shell-special characters
In my repository, I have files with names like "$list.xml." When I use "ga" to add those, it does not escape these characters.

```
$  gs
# On branch: scratch/test  |  [*] => $e*
#
➤ Changes not staged for commit
#
#       modified: [1] $list.xml
#
$  ga 1
fatal: pathspec '/Users/username/git/test/.xml' did not match any files
# Added '/Users/username/git/test/$list.xml'
```

If there's anything I can do to help fix this bug, please let me know. Thanks in advance!

### LINKED ISSUE #274: $e# variables are set incorrectly with `ll dir1 dir2`
```
% mkdir 1 2
% touch 1/file1 2/file2
% /bin/ls -lR
.:
total 0
drwxr-xr-x 1 ravi ravi 10 Sep 14 15:17 1
drwxr-xr-x 1 ravi ravi 10 Sep 14 15:17 2

./1:
total 0
-rw-r--r-- 1 ravi ravi 0 Sep 14 15:17 file1

./2:
total 0
-rw-r--r-- 1 ravi ravi 0 Sep 14 15:17 file2
% ll 1 2
-rw-r--r-- 1 ravi ravi 0 Sep 14 15:17 [1]  file1
-rw-r--r-- 1 ravi ravi 0 Sep 14 15:17 [2]  file2
% echo $e1
/home/ravi/code/scm_breeze/x/2/1:
% echo $e2
/home/ravi/code/scm_breeze/x/2/file1
%
```

### PR DESCRIPTION
**Pull request by HaleTom**
@ghthor and @nathanb could I have your feedback on this PR? Also, please tag anyone else you believe should review this PR.

**Fixes**

* Fix #167: "Empty string arguments not being passed to git"
* Fix #202: "Wrong argument quoting when filename uses spaces"
* Fix #204: "File names with shell-special characters"
* Fix #274: "$e# variables are set incorrectly with `ll dir1 dir2`"

**What's in the can:**

* Return (well, print actually) a quoted array from `scmb_expand_args()` via `declare -p`. This array can then be reconstituted in the caller by `eval "args=$(scmb_expand_args "$@")"`.
* Add `_safe_eval()`: quote `$@` elements before `eval` to prevent, eg, `eval "echo" "foo;date"` from running `date`. With `_safe_eval`, the output will be `foo;date`.
  * Add `_token_quote` test helper which prints the shell-escaped version of each of its arguments.
* Don't clobber user's `$IFS`: use locally scoped `$IFS` variables.
* Replace all instances of unquoted `$@` with `"$@"` to prevent splitting of array elements by $IFS
  * Likewise, replace `${array[@]}` with "${array[@]}"
* Invoke `readlink -f` in a way which can succeed
* Loosen regex match in `strip_colors()` to pass tests
  * Tests will otherwise fail with `$LS_COLOURS` as set by https://github.com/trapd00r/LS_COLORS
* `assert(Not)?Includes`: call `_includes` with an array
* `test_helper.sh`: Don't clobber global name/email config
* `shell_shortcuts_test.sh`: correct quoting
* `scmb_expand_args()`: return an array to fix quoting issues
* `_print_path`: remove unnecessary pipeline
* `test_ls_with_file_shortcuts()`: clean up temp files after use
* `ls_with_file_shortcuts()`: print error to STDERR
* Allow `run_tests.sh` to be executed from anywhere
* Ensure file ordering matches in both `ls` invocations in `ls_with_file_shortcuts`
* `git_show_affected_files:` make `$f` local

May fix #183: "Breaks cd when using other languages" (awaiting reproduction instructions)

May contain traces of peanuts. Reviews gratefully received.

### CODE REVIEW

**Comment by HaleTom** [comment]
Oops, I just read that it's a bad idea to `git push` with `-f` to a PR branch.

I did it at 2018-09-02 05:30 UTC if it helps to de-fuxor things at your end.

**Comment by HaleTom** [comment]
Awaiting "Upgrade Travis CI environment to newer versions of bash and zsh #264" to proceed.

**Comment by nathanb** [comment]
@HaleTom Thanks for the include, but I think you might be referring to @ndbroadbent instead.

**Comment by ghthor** [comment]
This PR is shaping up really nicely! I need to fix that Travis_CI build #264

**Comment by HaleTom** [comment]
This is finally green with Travis.

Can I please get a LGTM or two?

File: lib/git/aliases.sh [hunk 1]
```diff
@@ -17,7 +17,7 @@ unalias git > /dev/null 2>&1
 unset -f git > /dev/null 2>&1
 
 # Use the full path to git to avoid infinite loop with git function
-export _git_cmd="$(\which git)"
+export _git_cmd="$(bin_path git)"
```
Comment (jeffbyrnes) [inline lib/git/aliases.sh]: Alternatively, one could use `command -v git` instead.

File: lib/git/branch_shortcuts.sh [hunk 2]
```diff
@@ -32,14 +32,12 @@ EOF
 )"
 
   # Set numbered file shortcut in variable
-  local e=1
-  IFS=$'\n'
+  local e=1 IFS=$'\n'
```
Comment (jeffbyrnes) [inline lib/git/branch_shortcuts.sh]: I’m partial to keep things on separate lines, though declaring using `local` is definitely the right move.

File: lib/git/helpers.sh [hunk 3]
```diff
@@ -13,4 +13,11 @@ function fail_if_not_git_repo() {
...
-}
\ No newline at end of file
+}
+
+bin_path() {
+  if [[ -n ${ZSH_VERSION:-} ]];
+    then builtin whence -cp "$1" 2> /dev/null
+    else builtin type -P "$1"
+  fi
+}
```
Comment (jeffbyrnes) [inline lib/git/helpers.sh]: Aha! My previous comment seems moot, now that I see this.

File: lib/git/shell_shortcuts.sh [hunk 4]
```diff
@@ -23,19 +23,19 @@ if [ "$shell_command_wrapping_enabled" = "true" ] || [ "$bash_command_wrapping_e
...
+    for cmd in $scmb_wrapped_shell_commands; do
       if [ "${scmbDebug:-}" = "true" ]; then echo "SCMB: Wrapping $cmd..."; fi
 
       # Special check for 'cd', to make sure SCM Breeze is loaded after RVM
       if [ "$cmd" = 'cd' ]; then
         if [ -e "$HOME/.rvm" ] && ! type rvm > /dev/null 2>&1; then
-          echo -e "\033[0;31mSCM Breeze must be loaded \033[1;31mafter\033[0;31m RVM, otherwise there will be a conflict when RVM wraps the 'cd' command.\033[0m"
-          echo -e "\033[0;31mPlease move the line that loads SCM Breeze to the bottom of your ~/.bashrc\033[0m"
+          echo -e "\\033[0;31mSCM Breeze must be loaded \\033[1;31mafter\\033[0;31m RVM, otherwise there will be a conflict when RVM wraps the 'cd' command.\\033[0m"
+          echo -e "\\033[0;31mPlease move the line that loads SCM Breeze to the bottom of your ~/.bashrc\\033[0m"
```
Comment (jeffbyrnes) [inline lib/git/shell_shortcuts.sh]: Wouldn’t this cause those color codes to be output literally?

File: lib/git/aliases.sh (hunk 1 above)
Comment (HaleTom) [reply lib/git/aliases.sh]: Retrieve a binary pathname, not a function or alias.

File: lib/git/helpers.sh (hunk 3 above)
Comment (HaleTom) [reply lib/git/helpers.sh]: All good :)

File: lib/git/shell_shortcuts.sh (hunk 4 above)
Comment (HaleTom) [reply lib/git/shell_shortcuts.sh]: Nope.... `echo -e` wants a literal `\`, which is `"\\"` in double quotes.

**Comment by ghthor** [comment]
@HaleTom @jeffbyrnes Any more reservations? Can I get a LGTM for my conflict resolution in a8cf407?

**Pull request merged**